README.md
<p align="center">
    <a href="https://github.com/Flinesoft/AnyLint-Swift/actions?query=branch%3Amain">
        <img src="https://github.com/Flinesoft/AnyLint-Swift/workflows/CI/badge.svg"
            alt="CI">
    </a>
    <a href="https://github.com/Flinesoft/AnyLint-Swift/releases">
        <img src="https://img.shields.io/badge/Version-0.0.0-blue.svg"
             alt="Version: 0.0.0">
    </a>
    <a href="https://github.com/Flinesoft/AnyLint-Swift/blob/main/LICENSE">
        <img src="https://img.shields.io/badge/License-MIT-lightgrey.svg"
             alt="License: MIT">
    </a>
    <br />
    <a href="https://paypal.me/Dschee/5EUR">
        <img src="https://img.shields.io/badge/PayPal-Donate-orange.svg"
             alt="PayPal: Donate">
    </a>
    <a href="https://github.com/sponsors/Jeehut">
        <img src="https://img.shields.io/badge/GitHub-Become a sponsor-orange.svg"
             alt="GitHub: Become a sponsor">
    </a>
    <a href="https://patreon.com/Jeehut">
        <img src="https://img.shields.io/badge/Patreon-Become a patron-orange.svg"
             alt="Patreon: Become a patron">
    </a>
</p>

<p align="center">
  <a href="#swift">Swift</a>
  • <a href="#xcode">Xcode</a>
  • <a href="#docs">Docs</a>
  • <a href="#donation">Donation</a>
  • <a href="https://github.com/Flinesoft/AnyLint-Swift/issues">Issues</a>
  • <a href="#license">License</a>
</p>

# AnyLint-Swift

The [AnyLint](https://github.com/Flinesoft/AnyLint) community-collected configurations project for the Swift community.

## Configurations

A configuration can be placed into an AnyLint configuration file like so:

```Swift
// MARK: Swift/Core
try Lint.runChecks(source: .github(repo: "Flinesoft/AnyLint-Swift", version: "main", variant: "Swift/Core"))
```

### Swift

If you are using Swift, you will probably want to use the `Swift/Core` configuration.

Other configurations might follow and will be more opinionated towards a specific programming style.

### Xcode

There is no `Core` configuration here yet.

* If you have `ObjectiveC` files in your project, use the `Xcode/ObjectiveC` variant.
* If you have a projects with localized Strings files, use the `Xcode/Strings` variant.
* If you are using the [Carthage](https://github.com/Carthage/Carthage) dependency manager, use the `Xcode/Carthage` variant.
* If you are using [SwiftGen](https://github.com/SwiftGen/SwiftGen) for accessing resources, use the `Xcode/SwiftGen` variant.

Other configurations might follow and will be more opinionated towards a specific programming style.

### Docs

If you want to document your project, you will probably want to use the `Docs/Core` configuration.

Other configurations might follow and will be more opinionated towards a specific documentation styles.

## Donation

AnyLint-Swift was brought to you by [Cihat Gündüz](https://github.com/Jeehut) in his free time. If you want to thank me and support the development of this project, please **make a small donation on [PayPal](https://paypal.me/Dschee/5EUR)**. In case you also like my other [open source contributions](https://github.com/Flinesoft) and [articles](https://medium.com/@Jeehut), please consider motivating me by **becoming a sponsor on [GitHub](https://github.com/sponsors/Jeehut)** or a **patron on [Patreon](https://www.patreon.com/Jeehut)**.

Thank you very much for any donation, it really helps out a lot! 💯

## Contributing

Contributions are welcome. Feel free to open an issue on GitHub with your ideas or implement an idea yourself and post a pull request. If you want to contribute code, please try to follow the same syntax and semantic in your **commit messages** (see rationale [here](http://chris.beams.io/posts/git-commit/)). Also, please make sure to add an entry to the `CHANGELOG.md` file which explains your change.

## License

This library is released under the [MIT License](http://opensource.org/licenses/MIT). See LICENSE for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

<details>
<summary>Formatting Rules for Entries</summary>
Each entry should use the following format:

```markdown
- Summary of what was changed in a single line using past tense & followed by two whitespaces.  
  Issue: [#0](https://github.com/Flinesoft/AnyLint-Swift/issues/0) | PR: [#0](https://github.com/Flinesoft/AnyLint-Swift/pull/0) | Author: [Cihat Gündüz](https://github.com/Jeehut)
```

Note that at the end of the summary line, you need to add two whitespaces (`  `) for correct rendering on GitHub.

If needed, pluralize to `Tasks`, `PRs` or `Authors` and list multiple entries separated by `, `. Also, remove entries not needed in the second line.
</details>

## [Unreleased]
### Added
- None.
### Changed
- None.
### Deprecated
- None.
### Removed
- None.
### Fixed
- None.
### Security
- None.


Xcode/Strings.swift
#!/usr/local/bin/swift-sh
import AnyLint // @Flinesoft == wip/cg_template-system

try Lint.reportResultsToFile(arguments: CommandLine.arguments) {
    // MARK: - Variables
    let stringsFiles: Regex = #"^App/.*/\w+\.lproj/\w+\.strings$"#

    // MARK: - Checks
    // MARK: LocalizationWhitespaces
    try Lint.checkFileContents(
        checkInfo: "LocalizationWhitespaces: Do not add whitespaces to the beginning or end of localized strings – use String interpolation instead.",
        regex: #""\s*=\s*" +| +";"#,
        matchingExamples: [#""my.example.key1" = "(Total: ";\n"#, #""my.example.key2" = "\n    Title: ";\n"#, #""my.example.key3" = "    Title:";\n"#],
        nonMatchingExamples: [#""my.example.key1" = "(Total:";\n"#, #""my.example.key2" = "\n    Title:";\n"#, #""my.example.key3" = "Title:";\n"#],
        includeFilters: [stringsFiles]
    )
}


Xcode/Carthage.swift
#!/usr/local/bin/swift-sh
import AnyLint // @Flinesoft == wip/cg_template-system

try Lint.reportResultsToFile(arguments: CommandLine.arguments) {
    // MARK: - Variables
    let cartfile: Regex = #"^Cartfile$"#

    // MARK: - Checks
    // MARK: CartfileEmptyLines
    try Lint.checkFileContents(
        checkInfo: "CartfileEmptyLines: Make sure that each dependency is separated from the neighbors by an empty line.",
        regex: #"(#[^\n]+\n)?(git[^\n]+)\n([^\n]+)"#,
        matchingExamples: [
            """
            github "Flinesoft/HandySwift" ~> 3.1
            github "Flinesoft/HandyUIKit" ~> 1.9
            """,
            """
            # Handy Swift features that didn't make it into the Swift standard library.
            github "Flinesoft/HandySwift" ~> 3.1
            # Handy UI features that should have been part of UIKit in the first place.
            github "Flinesoft/HandyUIKit" ~> 1.9
            """
        ],
        nonMatchingExamples: [
            """
            github "Flinesoft/HandySwift" ~> 3.1
            github "Flinesoft/HandyUIKit" ~> 1.9
            """,
            """
            # Handy Swift features that didn't make it into the Swift standard library.
            github "Flinesoft/HandySwift" ~> 3.1
            # Handy UI features that should have been part of UIKit in the first place.
            github "Flinesoft/HandyUIKit" ~> 1.9
            """
        ],
        includeFilters: [cartfile],
        autoCorrectReplacement: "$1$2\n\n$3",
        autoCorrectExamples: [
            [
                "before": """
                    github "Flinesoft/HandySwift" ~> 3.1
                    github "Flinesoft/HandyUIKit" ~> 1.9
                    """,
                "after": """
                    github "Flinesoft/HandySwift" ~> 3.1
                    github "Flinesoft/HandyUIKit" ~> 1.9
                    """
            ],
            [
                "before": """
                    # Handy Swift features that didn't make it into the Swift standard library.
                    github "Flinesoft/HandySwift" ~> 3.1
                    # Handy UI features that should have been part of UIKit in the first place.
                    github "Flinesoft/HandyUIKit" ~> 1.9
                    """,
                "after": """
                    # Handy Swift features that didn't make it into the Swift standard library.
                    github "Flinesoft/HandySwift" ~> 3.1
                    # Handy UI features that should have been part of UIKit in the first place.
                    github "Flinesoft/HandyUIKit" ~> 1.9
                    """
            ],
        ]
    )

    // MARK: CartfileComments
    try Lint.checkFileContents(
        checkInfo: "CartfileComments: Make sure that each dependency has their GitHub tagline as a comment above it.",
        regex: #"(?<=\n\n)git[^\n]+"#,
        matchingExamples: [
            """
            github "Flinesoft/HandySwift" ~> 3.1
            github "Flinesoft/HandyUIKit" ~> 1.9
            """
        ],
        nonMatchingExamples: [
            """
            # Handy Swift features that didn't make it into the Swift standard library.
            github "Flinesoft/HandySwift" ~> 3.1
            # Handy UI features that should have been part of UIKit in the first place.
            github "Flinesoft/HandyUIKit" ~> 1.9
            """
        ],
        includeFilters: [cartfile]
    )

    // MARK: CartfileVersionSpecifier
    try Lint.checkFileContents(
        checkInfo: "CartfileVersionSpecifier: Make sure to allow for minor upgrades for all dependencies.",
        regex: #"(github "\w+\/\w+") *~> *(\d+\.\d+)\.\d+"#,
        matchingExamples: [#"github "Flinesoft/HandySwift" ~> 3.2.0"#],
        nonMatchingExamples: [#"github "Flinesoft/HandySwift" ~> 3.2"#, #"github "aromajoin/material-showcase-ios" ~> 0.7.1"#],
        includeFilters: [cartfile],
        autoCorrectReplacement: "$1 ~> $2",
        autoCorrectExamples: [
            ["before": #"github "Flinesoft/HandySwift" ~> 3.2.0"#, "after": #"github "Flinesoft/HandySwift" ~> 3.2"#],
        ]
    )
}


Xcode/SwiftGen.swift
#!/usr/local/bin/swift-sh
import AnyLint // @Flinesoft == wip/cg_template-system

try Lint.reportResultsToFile(arguments: CommandLine.arguments) {
    // MARK: - Variables
    let swiftFiles: Regex = #"^(App|Tests|UITests)/Sources/.*\.swift$"#
    let generatedSwiftFiles: Regex = #".*/Generated/.*\.swift|.*/SwiftGen/.*\.swift"#

    // MARK: - Checks
    // MARK: DynamicColorReference
    try Lint.checkFileContents(
        checkInfo: "DynamicColorReference: Don't use dynamic color references – use SwiftGen & Color instead.",
        regex: #"UIColor\(\s*named:\s*\""#,
        matchingExamples: [#"UIColor(named: "primary")"#],
        nonMatchingExamples: ["Colors.primary.color"],
        includeFilters: [swiftFiles],
        excludeFilters: [generatedSwiftFiles]
    )

    // MARK: DynamicStoryboardReference
    try Lint.checkFileContents(
        checkInfo: "DynamicStoryboardReference: Don't use dynamic storyboard references – use SwiftGen & StoryboardScene instead.",
        regex: #"UIStoryboard\(\s*name:\s*\""#,
        matchingExamples: [#"UIStoryboard(name: "LoginViewController")"#],
        nonMatchingExamples: ["StoryboardScene.Login.loginViewController.instantiate()"],
        includeFilters: [swiftFiles],
        excludeFilters: [generatedSwiftFiles]
    )

    // MARK: DynamicStringReference
    try Lint.checkFileContents(
        checkInfo: "DynamicStringReference@warning: Don't use dynamic localization string references via code strings – use SwiftGen & L10n instead.",
        regex: #"NSLocalizedString\s*\("#,
        matchingExamples: [#"NSLocalizedString(@"Test", comment: "");"#, #"NSLocalizedString("Test", comment: nil)"#],
        includeFilters: [swiftFiles],
        excludeFilters: [generatedSwiftFiles]
    )
}


Xcode/ObjectiveC.swift
#!/usr/local/bin/swift-sh
import AnyLint // @Flinesoft == wip/cg_template-system

try Lint.reportResultsToFile(arguments: CommandLine.arguments) {
    // MARK: - Variables
    let objectiveCFiles: Regex = #"^(App|Tests|UITests)/Sources/.*\.(h|m|mm)$"#

    // MARK: - Checks
    // MARK: BeforeClosingBracesWhitespace
    try Lint.checkFileContents(
        checkInfo: "BeforeClosingBracesWhitespace: Closing curly braces in ObjC should never be preceded by empty newlines.",
        regex: #"(?:\n *)+\n( *)\}"#,
        matchingExamples: ["\n        \n    }"],
        nonMatchingExamples: ["\n        return 5;\n    }"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "\n$1}",
        autoCorrectExamples: [
            ["before": "\n        \n    }", "after": "\n    }"],
            ["before": "\n        \n        \n    }", "after": "\n    }"],
        ]
    )

    // MARK: ElsePrefixWhitespace
    try Lint.checkFileContents(
        checkInfo: "ElsePrefixWhitespace: Else cases in ObjC should precede with a single whitespace like so: `} else {`.",
        regex: #"\}else|\} {2,}else"#,
        matchingExamples: ["}else {", "}  else {"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "} else",
        autoCorrectExamples: [
            ["before": "}else {", "after": "} else {"],
            ["before": "}  else {", "after": "} else {"],
        ]
    )

    // MARK: ElseSuffixWhitespace
    try Lint.checkFileContents(
        checkInfo: "ElseSuffixWhitespace: Else cases in ObjC should be followed with a whitespace like so: `} else {`.",
        regex: #"\} else\{|\} else\s{2,}\{|\} else\n\{"#,
        matchingExamples: ["} else{", "} else  {", "} else\n{", "} else\n  {"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "} else {",
        autoCorrectExamples: [
            ["before": "} else  {", "after": "} else {"],
            ["before": "} else{", "after": "} else {"],
        ]
    )

    // MARK: EmptyTodo
    try Lint.checkFileContents(
        checkInfo: "EmptyTodo: `// TODO:` comments should not be empty.",
        regex: #"// TODO: ?(\[[\d\-_a-z]+\])? *\n"#,
        matchingExamples: ["// TODO:\n", "// TODO: [2020-03-19]\n", "// TODO: [cg_2020-03-19]  \n"],
        nonMatchingExamples: ["// TODO: refactor", "// TODO: not yet implemented", "// TODO: [cg_2020-03-19] not yet implemented"],
        includeFilters: [swiftFiles]
    )

    // MARK: FunctionWhitespace
    try Lint.checkFileContents(
        checkInfo: "FunctionWhitespace: Whitespaces on ObjC functions should be of structure: `- (type)function:(Type *)param {`",
        regex: #"^([\+\-]) *\( *([^)]+) *\) *([^\n\{]+[^ \n])\s*\{\s*\n( *\S+)"#,
        matchingExamples: ["- (void)viewDidLoad\n{\n    blubb", "-(void)viewDidLoad {\n    blubb"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "$1 ($2)$3 {\n$4",
        autoCorrectExamples: [
            ["before": "- (void)viewDidLoad\n{\n    blubb", "after": "- (void)viewDidLoad {\n    blubb"],
            ["before": "-(void)viewDidLoad {\n    blubb", "after": "- (void)viewDidLoad {\n    blubb"],
        ]
    )

    // MARK: IfConditionWhitespace
    try Lint.checkFileContents(
        checkInfo: "IfConditionWhitespace: If conditions in ObjC should be surrounded with a whitespace like so: `if (condition) {`.",
        regex: #" if\s*\(\s*([^)]+[^\s])\s*\)\s*\{"#,
        matchingExamples: ["} else if (blubb == blubb){", "} else if(blubb == blubb) {", "} else if ( blubb == blubb) {"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: " if ($1) {",
        autoCorrectExamples: [
            ["before": "} else if (blubb == blubb){", "after": "} else if (blubb == blubb) {"],
            ["before": "} else if(blubb == blubb) {", "after": "} else if (blubb == blubb) {"],
            ["before": "} else if ( blubb == blubb) {", "after": "} else if (blubb == blubb) {"],
        ]
    )

    // MARK: MultilineWhitespaces
    try Lint.checkFileContents(
        checkInfo: "MultilineWhitespaces: Restrict whitespace lines to a maximum of one.",
        regex: #"\n( *\n){2,}"#,
        matchingExamples: ["}\n    \n     \n\nclass", "}\n\n\nvoid"],
        nonMatchingExamples: ["}\n    \n    class"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "\n\n",
        autoCorrectExamples: [
            ["before": "}\n    \n     \n\n    class", "after": "}\n\n    class"],
            ["before": "}\n\n\nvoid", "after": "}\n\nvoid"],
        ]
    )

    // MARK: TernaryOperatorWhitespace
    try Lint.checkFileContents(
        checkInfo: "TernaryOperatorWhitespace: There should be a single whitespace around each separator.",
        regex: #"(.*\S)\s*\?\s*(\w+)\s*:\s*(.*)"#,
        nonMatchingExamples: ["viewCtrl?.call(param: 50)"],
        includeFilters: [swiftFiles],
        autoCorrectReplacement: #"$1 ? $2 : $3"#,
        autoCorrectExamples: [
            ["before": "constant = singleUserMode() ? 0:28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ? 0 :28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ?0 : 28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ? 0: 28", "after": "constant = singleUserMode() ? 0 : 28"],
        ]
    )

    // MARK: TrailingWhitespaces
    try Lint.checkFileContents(
        checkInfo: "TrailingWhitespaces: There should be no trailing whitespaces in lines.",
        regex: #"([\S\n]) +\n"#,
        matchingExamples: ["}  \n", "\n    \n"],
        nonMatchingExamples: ["}\n    void"],
        includeFilters: [objectiveCFiles],
        autoCorrectReplacement: "$1\n",
        autoCorrectExamples: [
            ["before": "}  \n", "after": "}\n"],
            ["before": "\n    \n", "after": "\n\n"],
        ]
    )
}


Variants/sample.swift
#!/usr/local/bin/swift-sh
import AnyLint // @Flinesoft ~> 0.8.2

try Lint.reportResultsToFile(arguments: CommandLine.arguments) {
    // MARK: - Variables
    let swiftFiles: Regex = #"^(App|Tests|UITests)/Sources/.*\.swift$"#
    // TODO: [cg_2020-06-13] not checks implemented yet
    
    // MARK: - Checks
    // MARK: EmptyMethodBody
    try Lint.checkFileContents(
        checkInfo: "EmptyMethodBody: Don't use whitespace or newlines for the body of empty methods.",
        regex: ["declaration": #"(init|func [^\(\s]+)\([^{}]*\)"#, "spacing": #"\s*"#, "body": #"\{\s+\}"#],
        matchingExamples: [
            "init() { }",
            "init() {\n\n}",
            "init(\n    x: Int,\n    y: Int\n) { }",
            "func foo2bar()  { }",
            "func foo2bar(x: Int, y: Int)  { }",
            "func foo2bar(\n    x: Int,\n    y: Int\n) {\n    \n}",
        ],
        nonMatchingExamples: ["init() { /* comment */ }", "init() {}", "func foo2bar() {}", "func foo2bar(x: Int, y: Int) {}"],
        includeFilters: [swiftFiles],
        autoCorrectReplacement: "$declaration {}",
        autoCorrectExamples: [
            ["before": "init()  { }", "after": "init() {}"],
            ["before": "init(x: Int, y: Int)  { }", "after": "init(x: Int, y: Int) {}"],
            ["before": "init()\n{\n    \n}", "after": "init() {}"],
            ["before": "init(\n    x: Int,\n    y: Int\n) {\n    \n}", "after": "init(\n    x: Int,\n    y: Int\n) {}"],
            ["before": "func foo2bar()  { }", "after": "func foo2bar() {}"],
            ["before": "func foo2bar(x: Int, y: Int)  { }", "after": "func foo2bar(x: Int, y: Int) {}"],
            ["before": "func foo2bar()\n{\n    \n}", "after": "func foo2bar() {}"],
            ["before": "func foo2bar(\n    x: Int,\n    y: Int\n) {\n    \n}", "after": "func foo2bar(\n    x: Int,\n    y: Int\n) {}"],
        ]
    )

    // MARK: EmptyTodo
    try Lint.checkFileContents(
        checkInfo: "EmptyTodo: `// TODO:` comments should not be empty.",
        regex: #"// TODO: ?(\[[\d\-_a-z]+\])? *\n"#,
        matchingExamples: ["// TODO:\n", "// TODO: [2020-03-19]\n", "// TODO: [cg_2020-03-19]  \n"],
        nonMatchingExamples: ["// TODO: refactor", "// TODO: not yet implemented", "// TODO: [cg_2020-03-19] not yet implemented"],
        includeFilters: [swiftFiles, objectiveCFiles]
    )

    // MARK: EmptyType
    try Lint.checkFileContents(
        checkInfo: "EmptyType: Don't keep empty types in code without commenting inside why they are needed.",
        regex: #"(class|protocol|struct|enum) [^\{]+\{\s*\}"#,
        matchingExamples: ["class Foo {}", "enum Constants {\n    \n}", "struct MyViewModel(x: Int, y: Int, closure: () -> Void) {}"],
        nonMatchingExamples: ["class Foo { /* TODO: not yet implemented */ }", "func foo() {}", "init() {}", "enum Bar { case x, y }"],
        includeFilters: [swiftFiles]
    )
    
    // MARK: IfAsGuard
    try Lint.checkFileContents(
        checkInfo: "IfAsGuard: Don't use an if statement to just return – use guard for such cases instead.",
        regex: #" +if [^\{]+\{\s*return\s*[^\}]*\}(?! *else)"#,
        matchingExamples: [" if x == 5 { return }", " if x == 5 {\n    return nil\n}", " if x == 5 { return 500 }", " if x == 5 { return do(x: 500, y: 200) }"],
        nonMatchingExamples: [" if x == 5 {\n    let y = 200\n    return y\n}", " if x == 5 { someMethod(x: 500, y: 200) }", " if x == 500 { return } else {"],
        includeFilters: [swiftFiles]
    )
    
    // MARK: MultilineIfStructure
    try Lint.checkFileContents(
        checkInfo: "MultilineIfStructure: Make sure multiline if conditions are all on their own lines, indented by four spaces.",
        regex: #"^( *)if\s*(.*,) *\n((?:.+, *\n)*) *(.*\S)\s*\{\s*"#,
        includeFilters: [swiftFiles],
        autoCorrectReplacement: "$1if\n$1    $2\n$3$1    $4\n$1{\n$1    $5",
        autoCorrectExamples: [
            [
                "before": """
                        if let employee = UserDefaults.standard.currentEmployee(),
                            let employeeName = employeeName,
                            let employeeEmail = employeeEmail,
                            let recipientEmail = recipientEmail {
                                dataExporterHelper.exportData(
                    """,
                "after": """
                        if
                            let employee = UserDefaults.standard.currentEmployee(),
                            let employeeName = employeeName,
                            let employeeEmail = employeeEmail,
                            let recipientEmail = recipientEmail
                        {
                            dataExporterHelper.exportData(
                    """
            ],
            [
                "before": """
                        if let employee = UserDefaults.standard.currentEmployee(),
                            let recipientEmail = recipientEmail
                        {
                            dataExporterHelper.exportData(
                    """,
                "after": """
                        if
                            let employee = UserDefaults.standard.currentEmployee(),
                            let recipientEmail = recipientEmail
                        {
                            dataExporterHelper.exportData(
                    """
            ],
        ]
    )

    // MARK: MultilineGuardEnd
    try Lint.checkFileContents(
        checkInfo: "MultilineGuardEnd: Always close a multiline guard via `else {` on a new line indented like the opening `guard`.",
        regex: #"guard\s*([^\n]*,\n)+([^\n]*\S *)else\s*\{"#,
        matchingExamples: [
            """
            guard
                let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty() else {
                return
            }
            """,
            """
            guard let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty() else {
                return
            }
            """
        ],
        nonMatchingExamples: [
            """
            guard
                let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty()
            else {
                return
            }
            """,
            """
            guard let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty()
            else {
                return
            }
            """
        ],
        includeFilters: [swiftFiles]
    )

    // MARK: MultilineGuardStart
    try Lint.checkFileContents(
        checkInfo: "MultilineGuardStart: Always start a multiline guard via `guard` then a line break and all expressions indented.",
        regex: #"guard([^\n]*,\s*\n)+[^\n]*\s*else\s*\{"#,
        matchingExamples: [
            """
            guard let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty()
            else {
                return
            }
            """,
            """
            guard let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty() else {
                return
            }
            """
        ],
        nonMatchingExamples: [
            """
            guard
                let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty()
            else {
                return
            }
            """,
            """
            guard
                let collection = viewModel.myCollection(),
                !collection.compactMap({ OtherType($0) }).isEmpty() else {
                return
            }
            """
        ],
        includeFilters: [swiftFiles]
    )
    
    // MARK: NilCoalescingOperator
    try Lint.checkFileContents(
        checkInfo: "NilCoalescingOperator: Prefer nil coalescing operator over `variable != nil ? variable! : alternative`.",
        regex: #"(\w+)\s*!=\s*nil\s*\?\s*\1!\s*:\s*(.*)"#,
        includeFilters: [swiftFiles],
        autoCorrectReplacement: "$1 ?? $2",
        autoCorrectExamples: [
            [
                "before": "    let message = errorMessage != nil ? errorMessage! : L10n.Global.Info.success\n",
                "after": "    let message = errorMessage ?? L10n.Global.Info.success\n",
            ],
            [
                "before": "param: callFunction(errorMessage != nil ? errorMessage! : L10n.Global.Info.success),\n",
                "after": "param: callFunction(errorMessage ?? L10n.Global.Info.success),\n",
            ],
        ]
    )
    
    // MARK: SingleLineGuard
    try Lint.checkFileContents(
        checkInfo: "SingleLineGuard: Use a single line guard for simple checks.",
        regex: #"guard\s*([^\{]{2,80})\s+else\s*\{\s*\n\s*(return[^\n]{0,40}|continue|fatalError\([^\n;]+\))\s*\}"#,
        matchingExamples: ["guard x else {\n  return\n}", "guard x else {\n  return 2 * x.squared(x: {15})\n}", #"guard x else {\#n  fatalError("some message: \(x)")\#n}"#],
        nonMatchingExamples: ["guard x else { return }", "guard x else { return 2 * x.squared(x: {15}) }", #"guard x else { fatalError("some message: \(x)") }"#],
        includeFilters: [swiftFiles],
        autoCorrectReplacement: #"guard $1 else { $2 }"#,
        autoCorrectExamples: [
            ["before": "guard let x = y?.x(z: 5) else {\n  return\n}", "after": "guard let x = y?.x(z: 5) else { return }"],
            ["before": "guard let x = y?.x(z: 5) else {\n  return 2 * x.squared(x: {15})\n}", "after": "guard let x = y?.x(z: 5) else { return 2 * x.squared(x: {15}) }"],
            ["before": "guard let x = y?.x(z: 5)\nelse {\n  return\n}", "after": "guard let x = y?.x(z: 5) else { return }"],
        ]
    )
    
    // MARK: SingletonDefaultPrivateInit
    try Lint.checkFileContents(
        checkInfo: "SingletonDefaultPrivateInit: Singletons with a `default` object (pseudo-singletons) should not declare init methods as private.",
        regex: #"class +(?<TYPE>\w+)(?:<[^\>]+>)? *\{.*static let `default`(?:: *\k<TYPE>)? *= *\k<TYPE>\(.*(?<=private) init\(\m"#,
        matchingExamples: ["class MySingleton {\n  static let `default` = MySingleton()\n\n  private init() {}\n"],
        nonMatchingExamples: ["class MySingleton {\n  static let `default` = MySingleton()\n\n  init() {}\n"],
        includeFilters: [swiftFiles]
    )

    // MARK: SingletonSharedFinal
    try Lint.checkFileContents(
        checkInfo: "SingletonSharedFinal: Singletons with a single object (`shared`) should be marked as final.",
        regex: #"(?<!final )class +(?<TYPE>\w+)(?:<[^\>]+>)? *\{.*static let shared(?:: *\k<TYPE>)? *= *\k<TYPE>\(\m"#,
        matchingExamples: ["\nclass MySingleton {\n  static let shared = MySingleton()\n\n  private init() {}\n"],
        nonMatchingExamples: ["\nfinal class MySingleton {\n  static let shared = MySingleton()\n\n  private init() {}\n"],
        includeFilters: [swiftFiles]
    )

    // MARK: SingletonSharedPrivateInit
    try Lint.checkFileContents(
        checkInfo: "SingletonSharedPrivateInit: Singletons with a single object (`shared`) should declare their init method(s) as private.",
        regex: #"class +(?<TYPE>\w+)(?:<[^\>]+>)? *\{.*static let shared(?:: *\k<TYPE>)? *= *\k<TYPE>\(.*(?<= |\t|public|internal) init\(\m"#,
        matchingExamples: ["\nfinal class MySingleton {\n  static let shared = MySingleton()\n\n  init() {}\n"],
        nonMatchingExamples: ["\nfinal class MySingleton {\n  static let shared = MySingleton()\n\n  private init() {}\n"],
        includeFilters: [swiftFiles]
    )

    // MARK: SingletonSharedSingleObject
    try Lint.checkFileContents(
        checkInfo: "SingletonSharedSingleObject: Singletons with a `shared` object (real Singletons) should not have other static let properties. Use `default` instead (if needed).",
        regex: #"class +(?<TYPE>\w+)(?:<[^\>]+>)? *\{.*(?:static let shared(?:: *\k<TYPE>)? *= *\k<TYPE>\(.*static let \w+(?:: *\k<TYPE>)? *= *\k<TYPE>\(|static let \w+(?:: *\k<TYPE>)? *= *\k<TYPE>\(.*static let shared(?:: *\k<TYPE>)? *= *\k<TYPE>\()\m"#,
        matchingExamples: ["\nfinal class MySingleton {\n  static let shared = MySingleton(url: productionUrl)\n  static let test = MySingleton(url: testUrl)"],
        nonMatchingExamples: ["\nfinal class MySingleton {\n  static let shared = MySingleton()\n\n  private init() {}\n"],
        includeFilters: [swiftFiles]
    )

    // MARK: SwitchAssociatedValueStyle
    try Lint.checkFileContents(
        checkInfo: "SwitchAssociatedValueStyle: Always put the `let` in front of case – even if only one associated value captured.",
        regex: #"case +[^\(][^\n]*(\(let |[^\)], let)"#,
        matchingExamples: ["case .addText(let text, let font, let textColor):", "case .addImage(let image)"],
        nonMatchingExamples: ["case let .addText(text, font, textColor):", "case let .addImage(image)"],
        includeFilters: [swiftFiles]
    )

    // MARK: TernaryOperatorWhitespace
    try Lint.checkFileContents(
        checkInfo: "TernaryOperatorWhitespace: There should be a single whitespace around each separator.",
        regex: #"(.*\S)\s*\?\s*(\w+)\s*:\s*(.*)"#,
        nonMatchingExamples: ["viewCtrl?.call(param: 50)"],
        includeFilters: [swiftFiles, objectiveCFiles],
        autoCorrectReplacement: #"$1 ? $2 : $3"#,
        autoCorrectExamples: [
            ["before": "constant = singleUserMode() ? 0:28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ? 0 :28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ?0 : 28", "after": "constant = singleUserMode() ? 0 : 28"],
            ["before": "constant = singleUserMode() ? 0: 28", "after": "constant = singleUserMode() ? 0 : 28"],
        ]
    )
    
    // MARK: TupleIndex
    try Lint.checkFileContents(
        checkInfo: "TupleIndex: Prevent unwrapping tuples by their index – define a typealias with named components instead.",
        regex: #"(\$\d|\w*[^%\d \(\[\{])\.\d"#,
        matchingExamples: ["$0.0", "$1.2", "tuple.0"],
        nonMatchingExamples: ["$0.key", "tuple.key", "%1$d of %2$d"],
        includeFilters: [swiftFiles]
    )

    // MARK: UnnecessaryNilAssignment
    try Lint.checkFileContents(
        checkInfo: "UnnecessaryNilAssignment: Don't assign nil as a value when defining an optional type – it's nil by default.",
        regex: #"(\svar +\w+\s*:\s*[^\n=]+\?)\s*=\s*nil"#,
        matchingExamples: ["class MyClass {\n  var count: Int? = nil\n", "class MyClass {\n  var dict: Dictionary<String, AnyObject>? = nil\n"],
        nonMatchingExamples: ["class MyClass {\n  let count: Int? = nil\n", "funct sum(dict: Dictionary<String, AnyObject>? = nil,"],
        includeFilters: [swiftFiles],
        autoCorrectReplacement: "$1",
        autoCorrectExamples: [
            ["before": "class MyClass {\n  var  count:Int?=nil\n", "after": "class MyClass {\n  var  count:Int?\n"],
            ["before": "class MyClass {\n  var dict: Dictionary<String, AnyObject>? = nil\n", "after": "class MyClass {\n  var dict: Dictionary<String, AnyObject>?\n"],
        ]
    )
    
    // MARK: ViewControllerVariableNaming
    try Lint.checkFileContents(
        checkInfo: "ViewControllerVariableNaming: Always name your view controller variables with the suffix `ViewCtrl` or just `viewCtrl`.",
        regex: #"(var|let) +(vc|viewcontroller|viewc|vcontroller)[ :][^\n=]*=\i"#,
        matchingExamples: ["let vc =", "var viewController =", "let viewc =", "let vcontroller =", "var vc: MyViewController =", "let viewController: MyViewController<T> = "],
        nonMatchingExamples: ["let viewCtrl =", "let viewCtrl: MyViewController =", "var myViewCtrl: MyViewController<T> ="],
        includeFilters: [swiftFiles]
    )
}
